### PR TITLE
Correct update count with Gluon trainer and update_on_kvstore=False

### LINKED ISSUE #12713: distributed kvstore bug in MXNet
I'm using distributed kvstore with Gluon trainer. I found the two following bugs:

1. Initializing `trainer = gluon.Trainer(update_on_kvstore=True)` doesn't work. Inspecting `trainer._update_on_kvstore` shows that the value is still set to `False`. 

2. When distributed kvstore is used, by default `gluon.Trainer` doesn't work with `mx.optimizer.LRScheduler` if a worker has more than 1 GPU. To be more specific, the trainer updates once per GPU, the `LRScheduler` object is shared across GPUs and get a wrong update count. 

This means one cannot train imagenet classification using resnet with gluon trainer.

### LINKED ISSUE #13752: Adam, AdaMax and FTML cannot be used with Trainer(update_on_kv=False)
These optimizers adjust the learning rate by the number of optimization steps, which is problematic if `update_on_kv=False` is set and multiple GPUs share the same optimizer object (leading to incorrect count of optimization steps). For example: https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/optimizer/optimizer.py#L1093

cc @szha @sandeep-krishnamurthy

### PR DESCRIPTION
**Pull request by ptrendx**
## Description ##
This PR fixes the update count when Gluon trainer is created with multiple devices and `update_on_kvstore=False`.
Fixes #13752, fixes #12713

@eric-haibin-lin This slightly hacky but should work for all cases. Thoughts?

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
- The real fix here should be moving the _index_update_count to Updater (which is per device) from optimizer (which is common to all updaters), but that would require breaking API change. FYI @szha

### CODE REVIEW

**Comment by karan6181** [comment]
@mxnet-label-bot add [Bug, Gluon, Optimizer, pr-awaiting-review]

**Comment by karan6181** [comment]
@ptrendx Thank you for the contribution!

@eric-haibin-lin Can you please review this PR?

File: python/mxnet/optimizer/optimizer.py [hunk 1]
```diff
@@ -380,6 +381,11 @@ def set_wd_mult(self, args_wd_mult):
                     self.wd_mult[name] = float(attr[name]['__wd_mult__'])
         self.wd_mult.update(args_wd_mult)
 
+    def set_current_context(self, ctx):
+        if ctx not in self._all_index_update_counts:
```
Comment (szha) [inline python/mxnet/optimizer/optimizer.py]: rename to _set_current_context, and document the type of ctx?

Comment (ptrendx) [reply python/mxnet/optimizer/optimizer.py]: Will do

Comment (ptrendx) [reply python/mxnet/optimizer/optimizer.py]: Done.
